src/include/btmem.h

```diff
@@ -616,17 +616,41 @@ struct __wt_page {
  * but it's not always required: for example, if a page is locked for splitting, or being created or
  * destroyed.
  */
-#define WT_INTL_INDEX_GET_SAFE(page) __wt_atomic_load_pointer(&(page)->u.intl.__index)
+#ifdef TSAN_BUILD
+/*
+ * TSan doesn't detect the acquire/release barriers used in our normal WT_INTL_INDEX_* functions, so
+ * use __atomic intrinsics instead. We can use __atomics here as MSVC doesn't support TSan.
+ */
+#define WT_INTL_INDEX_GET_SAFE(page, pindex)                                   \
+    do {                                                                       \
+        (pindex) = __atomic_load_n(&(page)->u.intl.__index, __ATOMIC_ACQUIRE); \
+    } while (0)
+#define WT_INTL_INDEX_GET(session, page, pindex)                          \
+    do {                                                                  \
+        WT_ASSERT(session, __wt_session_gen(session, WT_GEN_SPLIT) != 0); \
+        WT_INTL_INDEX_GET_SAFE(page, (pindex));                           \
+    } while (0)
+#define WT_INTL_INDEX_SET(page, v)                                        \
+    do {                                                                  \
+        __atomic_store_n(&(page)->u.intl.__index, (v), __ATOMIC_RELEASE); \
+    } while (0)
+#else
+#define WT_INTL_INDEX_GET_SAFE(page, pindex)                          \
+    do {                                                              \
+        WT_ACQUIRE_BARRIER();                                         \
+        (pindex) = __wt_atomic_load_pointer(&(page)->u.intl.__index); \
```

Comment:
This should be a WT_ACQUIRE_READ().

Reply:
Updated in [399fa38](https://github.com/wiredtiger/wiredtiger/pull/10707/commits/399fa3821e03e50e02fe9657412386c6f58d4f74). I'd actually put the barrier on the wrong side of the instruction in the original commit.

One thing to call out here is `WT_INTL_INDEX_GET_SAFE` will now use `ldapr` instructions on compatible platforms. We're getting acquire semantics so I assume we maintain the atomicity guarantees of the original `__wt_atomic_load_pointer`.

Reply:
> One thing to call out here is WT_INTL_INDEX_GET_SAFE will now use ldapr instructions on compatible platforms

Should give the guarantees we want. We're going from a raw load from memory to a RCpc load.